idaes/core/util/unit_costing.py

```diff
@@ -905,61 +965,201 @@ def p_factor_rule(self):
+                12.32*(Q/pyunits.BTU*pyunits.hr)**0.64
         elif fired_type == 'dowtherm_a':
-            return self.base_cost == 12.74*(Q/pyunits.BTU*pyunits.hr)**0.65
+            return self.base_cost_per_unit ==\
+                12.74*(Q/pyunits.BTU*pyunits.hr)**0.65
         elif fired_type == 'steam_boiler':
-            return self.base_cost == 0.367*(Q/pyunits.BTU*pyunits.hr)**0.77
+            return self.base_cost_per_unit ==\
+                0.367*(Q/pyunits.BTU*pyunits.hr)**0.77
         else:
             raise Exception('fired heater type not supported')
-    self.cb_eq = Constraint(rule=CB_rule)
+    self.base_cost_per_unit_eq = Constraint(rule=CB_rule)
 
-    # # base cost fired heater
+    @self.Expression(doc="Base cost for all units installed")
+    def base_cost(self):
+        return self.base_cost_per_unit * self.number_of_units
+
+    # purchase cost fired heater
     def CP_rule(self):
         return self.purchase_cost == self.material_factor \
             * (self.parent_block().parent_block().costing.CE_index/500) \
             * self.pressure_factor * self.base_cost
-    self.cp_eq = Constraint(rule=CP_rule)
+    self.cp_cost_eq = Constraint(rule=CP_rule)
 
 
-def cstr_costing(self, alignment, Mat_factor, weight_limit, L_D_range, PL):
+def cstr_costing(self, alignment='vertical', Mat_factor='carbon_steel',
+                 weight_limit='option1', L_D_range='option1', PL=True):
     vessel_costing(self, alignment=alignment,
                    Mat_factor=Mat_factor,
                    weight_limit=weight_limit,
                    L_D_range=L_D_range,
                    PL=PL)
 
 
-def flash_costing(self, alignment, Mat_factor, weight_limit, L_D_range, PL):
+def flash_costing(self, alignment='vertical', Mat_factor='carbon_steel',
+                  weight_limit='option1', L_D_range='option1', PL=True):
     vessel_costing(self, alignment=alignment,
                    Mat_factor=Mat_factor,
                    weight_limit=weight_limit,
                    L_D_range=L_D_range,
                    PL=PL)
 
 
-def rstoic_costing(self, alignment, Mat_factor, weight_limit, L_D_range, PL):
+def rstoic_costing(self, alignment='vertical', Mat_factor='carbon_steel',
+                    weight_limit='option1', L_D_range='option1', PL=True):
     vessel_costing(self, alignment=alignment,
                    Mat_factor=Mat_factor,
                    weight_limit=weight_limit,
                    L_D_range=L_D_range,
                    PL=PL)
 
 
-def pfr_costing(self, alignment, Mat_factor, weight_limit, L_D_range, PL):
+def pfr_costing(self, alignment='horizontal', Mat_factor='carbon_steel',
+                    weight_limit='option1', L_D_range='option1', PL=True):
     vessel_costing(self, alignment=alignment,
                    Mat_factor=Mat_factor,
                    weight_limit=weight_limit,
                    L_D_range=L_D_range,
                    PL=PL)
+
+
+def initialize(self):
```

Comment:
This PR generally looks good to me.  I don't know too much about the costing framework and why it is implemented the way it is, but looking at this initialize method makes me want to raise this question again.  Miguel and I talked about this a little, so I'm not going to insist on a different implementation.  I'm just curious about really.

There are a lot of if hasattrs in this.  I would think the normal why to do this would be to have the costing block class be derived from ProcessBase, then build up the initialize methods though inheritance.   Like there is a common costing block, then from that derive one for pumps, from that specific pumps.  Naively that's how I would have done this.

Reply:
This also addresses (indirectly) to @andrewlee94's import comment.

Reply:
I did ask the same question, but we left it this way for now due to time constraints. However, I agree that going forward we should change this to inherit from ProcessBase and have different classes for each type of unit operation, each with their own initialize method.